### DISCUSSION TITLE
[help] How can I use the dynamic target naming functions for a targetopia package?

### DISCUSSION THREAD

**Discussion by billdenney** [ropensci/targets#1173] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I'm hoping to push `nlmixr2targets` to CRAN in the next few weeks.  One feature that I want to get in before sending it to CRAN is a way to allow a user to generate many models as a list easily.  The many models would all use the same source data, and they would have long descriptions.  A call to this may look something like:

```
tar_nlmixr2_multi(
  all_models,
  data = my_data,
  est = "focei",
  "Linear model" = mod_linear,
  "Emax model" = mod_emax
)
```

As an internal step, it would do something like the following where `93427bac` and `89053bd3` are arbitrary hashes.  (I realize that the code in the `nlmixr2targets` package would not look directly like this as there would be extensive use of `substitute()` and similar to make the names.  Hopefully the example is clear.)

```
all_models_93427bac <-
  nlmixr2(
    object = mod_linear,
    data = my_data,
    est = "focei"
  )
all_models_89053bd3 <-
  nlmixr2(
    object = mod_emax,
    data = my_data,
    est = "focei"
  )
all_models <-
  list(
    "Linear model" = all_models_93427bac,
    "Emax model" = all_models_89053bd3
  )
```

In essence, I want to use dynamic branching, I think.  But, I don't want to reinvent the wheel.  Is there a way that I can use the same methods that `targets` uses for creating dynamic branching names within the `nlmixr2targets` package?

**Comment by wlandau** [comment]
If you use dynamic branching, then the names are created automatically. But if the number of models is not in the thousands, and if those models can have informative names, then static branching could work well. Target factories in `stantargets` and `jagstargets` use static branching to run multiple model files on the same dataset. The relevant code for `tar_stan_mcmc()` is in https://github.com/ropensci/stantargets/blob/main/R/tar_stan_mcmc.R and https://github.com/ropensci/stantargets/blob/main/R/utils_targets.R.

**Reply by billdenney** [reply]
The typical workflow will have ~tens of models.  Creating informative names is not the goal here because I don't want to ask the user for an informative name that would work as an R object, and I want to allow the user to change the description (like `"Linear model"`) without re-running the model.  So, I don't want to make the name dependent on the description.

I can auto-generate the name from the target name (`all_models`) and a digest of the `object` argument. But, if there were a way to autogenerate the name within `targets`, I'd prefer to use that.

Based on your response, it sounds like there's not a standard way to generate a hash name.  Is that right?

**Reply by wlandau** [reply]
> I can auto-generate the name from the target name (all_models) and a digest of the object argument. 

Yeah, that sounds like a reasonable idea. Assuming all the objects upstream of the models are deterministic and unique, you can take `digest(object)` to generate a suffix for each name. 

> But, if there were a way to autogenerate the name within targets, I'd prefer to use that.

This kind of use case isn't something `targets` can realistically anticipate or automate. It's a matter of finding out what makes sense to hash to generate a name.
